src/typecheck.rs

```diff
@@ -1,11 +1,1384 @@
+            let inner_type = typeck_expr(inner, ctx)?;
+
+            // According to the [T-Cast] rule, both the inner expression and the target type
+            // must be numeric types with bounds
+            match (&inner_type, typ) {
+                (Type::NumT(_), Type::NumT(_)) => {
+                    // The cast is valid, return the target type
+                    Ok(typ.clone())
+                }
+                (_, Type::NumT(_)) => {
+                    // The inner expression is not a numeric type
+                    Err(TypeError::TypeMismatch {
+                        expected: Type::NumT(0..1), // Example numeric type
+                        actual: inner_type,
+                    })
+                }
+                (Type::NumT(_), _) => {
+                    // The target type is not a numeric type
+                    Err(TypeError::TypeMismatch {
+                        expected: Type::NumT(0..1), // Example numeric type
+                        actual: typ.clone(),
+                    })
+                }
+                _ => {
+                    // Neither the inner expression nor the target type are numeric types
+                    Err(TypeError::TypeMismatch {
+                        expected: Type::NumT(0..1), // Example numeric type
+                        actual: inner_type,
+                    })
+                }
+            }
+        }
+        // Call expressions are always of the type of the function
+        Expr::Call { callee, args } => {
+            // Type check each argument
+            let mut arg_types: Vec<Type> = Vec::new();
+            for arg in args {
+                arg_types.push(typeck_expr(arg, ctx)?);
+            }
+
+            // Look up the function in the function environment
+            let function = ctx
+                .funcs
+                .get(callee)
+                .ok_or(TypeError::UndefinedFunction(*callee))?;
+
+            // Check that the number of arguments matches the number of parameters
+            if arg_types.len() != function.params.len() {
+                return Err(TypeError::WrongNumberOfArguments {
+                    expected: function.params.len(),
+                    actual: arg_types.len(),
+                });
+            }
+
+            // Verify that argument types match parameter types
+            for (arg_type, (_, param_type)) in arg_types.iter().zip(function.params.iter()) {
+                if arg_type != param_type {
+                    return Err(TypeError::TypeMismatch {
+                        expected: param_type.clone(),
+                        actual: arg_type.clone(),
+                    });
+                }
+            }
+
+            // Return the function's return type
+            Ok(function.ret_typ.clone())
+        }
+        // Pattern matching is always of the type of the scrutinee
+        Expr::Match { scrutinee, cases } => {
+            // Type check the scrutinee
```

Comment:
ditto for these comments, this applies universally: if a comment is just explaining what the code does (as opposed to why, or explaining a subtle edge case), get rid of it. they get in the way of readability of the code.